### ISSUE TITLE
App interface enhancements

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#965]
In working on how to launch a CyberGIS app from HydroShare with @selimnairb and @ahmetartu we reviewed the app calling protocol and identified the following deficiencies/concerns.

An example app launch call is https://apps.hydroshare.org/apps/raster-viewer/?res_id=877bf9ed9e66468cadddb229838a9ced&usr=dtarb&src=hs
1. The tag res_id is limiting as the target system may have this used for another purpose.  It would be better for the calling tag to be user specified as part of the app instance setup.
2. There are concerns over trust in the query string.  Specifically passing of usr= would be better handled by relying on OAuth and passing of src= would be better handled using the refer header.
3.  There does not seem to be a way to test apps from a development platform.  I tried to test an app from park.hydroshare.org, but got the error that there was no such resource in HydroShare.  I suggest that at least appsdev.hydroshare.org be configured to take inputs from our development platforms for testing.

@selimnairb and @ahmetartu feel free to elaborate or correct anything I may have not got right from our discussion.  @zhiyuli @danames please ask @selimnairb and @ahmetartu if you have questions.

**Comment by selimnairb** [comment]
Here is a proposed architecture for orienting apps around content types rather than resources:

- Currently, an App resource can be applied to one or more resource types.
- However for some resource types, e.g. the soon-to-be-written RHESSys model instance, there can be multiple data files of different types within one resource (e.g. a DEM raster, a landcover raster; climate forcing timeseries, model output timeseries)
- To allow generic tools (i.e. Apps) to be applied to these multiple data sets within a resource, we should move HydroShare apps from having n resource focus to being focused on content types.
- Each resource type would advertise: (1) zero or more content types that it is capable of supplying (e.g. through a get_supported_content_types() method on AbstractResource, and overridden by concrete resource types); and (2) for each content type, file paths within the resource that are instances of the content type (e.g. through a get_files_of_content_type(content_type) method on AbstractResource, and overridden by concrete resource types)
- A given App resource would be associated with one or more “supported content types” (instead of “supported resource types”).  Then, at runtime, when an App is being applied to a given resource instance, it would call get_files_of_content_type() for each content type that the App knows how to deal with.
- In the case of resource type instances that only supply one dataset of one content type, the App would behave the same as it does now.  However, for resource type instances that support multiple datasets or content types, then the App could allow the user to choose what dataset(s) of each content type to be displayed/visualized/whatever.
- I think that we can relatively easily create the concept of named content types that consist of one or more MIME types.  For example, a GeographicRaster would contain a list of MIME types for raster formats supported by our version of GDAL.  We can create these lists in the admin console.

**Comment by zhiyuli** [comment]
@selimnairb 
as for the RHESSys resource type, what should the get_supported_content_types() method return?
and what should it return for the Raster resource type?

**Comment by selimnairb** [comment]
@zhiyuli First, we will need to define content types supported by HydroShare; a content type is a named set of one or more MIME types.  We can start by creating content types supported by each of the current resource types.  For example, a "geographic raster" content type (not to be confused with a geographic raster resource), could be a list of MIME types supported by the as-compiled version of GDAL HydroShare uses; a "point timeseries" content type could be MIME types associated with WaterML, CSV, an sqlite database, etc.  

For the RHESSys model instance, get_supported_content_types() would return: {'geographic_raster', 'point_timeseries', ...}.  Likewise, the raster resource type get_supported_content_types() would return {'geographic_raster'}.

Then get_files_of_content_type('geographic_raster') called on an instance of a raster resource would return the content file that represents the geographic raster (e.g. data/contents/BearCk.tif for a GeoTIFF).  For a RHESSys model instance, this would be more complicated, as there can be more than one raster, and these rasters can be in different locations.  It would be up to the author of the RHESSys model instance type to override get_files_of_content_type().  However the base/abstract class could have a default implementation that simply iterates over all resource files, returning files that match the desired content type (sub-classes would override this if they wanted to restrict the list of files returned, or had a more efficient way, e.g. based on resource type-specific metadata, of returning the list of files).  This way, resource type authors would not need to do additional work unless they had special needs.

**Comment by danames** [comment]
Let's assign some time at the all hands meeting to discuss this. I don't
want to add a lot of complexity at this stage but see the value in
extending apps to work on content rather than resource type.

**Comment by selimnairb** [comment]
@danames Agreed that we don't want to add too much complexity and development cost.  I believe this can be done relatively simply as outlined above, with the majority of the code in the BaseResource class.  We should discuss this at a break-out group at the all-hands meeting so that we can make sure there are no special cases that add complexity that we cannot afford.  If we agree to proceed, we may be able to get this done during the hackathon portion of the meeting.

**Comment by selimnairb** [comment]
@alvacouch Please comment on this issue.

**Comment by zhiyuli** [comment]
One concern: HydroShare provides detailed metadata for each resource type. Apps can use these metadata directly instead of exacting them from the resource files by themselves. 

If a "Composite" resource type has rasters, vectors, netcdf and timeseries, will each of their metadata be extracted by hydroshare?

**Comment by zhiyuli** [comment]
The above concern also applies to issue #993 which is being addressed by PR #1003.
When launching an app, sometimes user needs to pass some resource-specific metadata in url query string. These metadata is currently retrieved from the requested resource obj  --> metadata obj.

Say we launch an raster viewer app from a "composite resource" (holds raster, vector and timeseries) landing page. The raster viewer app only uses raster resource files. If the "Composite" resource type cannot provide metadata for this raster file, current implementation for "making app url more flexible" will not work any more.

**Comment by selimnairb** [comment]
@zhiyuli Is a composite resource a Collection Resource (a resource that contains other resources), or is it merely a resource that contains a bunch of files?  

In either case, as all resource types will support the *get_supported_content_types()* and *get_files_of_content_type(content_type)*, an App resource instance can discover what it needs by calling these functions on the object resource.  In the case of Collection Resources, the calls can be broadcast to all contained resources, and the results collated before being returned to the App.

**Comment by zhiyuli** [comment]
@selimnairb I mean a resource that contains a bunch of files. So this resource will extract metadata for each file (raster, vector and ...) it holds?

**Comment by horsburgh** [comment]
@zhiyuli - The resource metadata is for the resource as a whole. So, unless we design the "composite" resource so it contains file-specific metadata for every content file, there wouldn't be specific metadata for each content file. We should discuss whether we want to have file-specific metadata (and what it should be) for each content file in a "composite" resource.

For "collection" resources, since they are just pointers to other resources within HydroShare, there would be both the resource level metadata for the "collection" resource AND resource level metadata for every collected resource (because they are all separate resources and have their own metadata).

**Comment by zhiyuli** [comment]
@horsburgh Thanks for this clarification. I am more concerned about the topic in your first paragraph, especially when a "composite" resource contains multiple resource files of different types (raster, vector and...).

**Comment by selimnairb** [comment]
@zhiyuli @horsburgh Reproduced from above:
> Each resource type would advertise: (1) zero or more content types that it is capable of supplying (e.g. through a get_supported_content_types() method on AbstractResource, and overridden by concrete resource types); and (2) for each content type, file paths within the resource that are instances of the content type (e.g. through a get_files_of_content_type(content_type) method on AbstractResource, and overridden by concrete resource types)

And:
> Then get_files_of_content_type('geographic_raster') called on an instance of a raster resource would return the content file that represents the geographic raster (e.g. data/contents/BearCk.tif for a GeoTIFF). For a RHESSys model instance, this would be more complicated, as there can be more than one raster, and these rasters can be in different locations. It would be up to the author of the RHESSys model instance type to override get_files_of_content_type(). However the base/abstract class could have a default implementation that simply iterates over all resource files, returning files that match the desired content type (sub-classes would override this if they wanted to restrict the list of files returned, or had a more efficient way, e.g. based on resource type-specific metadata, of returning the list of files). This way, resource type authors would not need to do additional work unless they had special needs.

**Comment by horsburgh** [comment]
@selimnairb - these are great ideas, and we definitely need ways of getting at the individual files of a composite resource. 

I think @zhiyuli is wondering if we will go a step further than providing access to the individual files to providing file-specific metadata. For example - if you had a time series in a composite resource, could you access all of the time series metadata?  Or, if you had a raster in the composite resource, can you get the extents of the raster and other raster specific metadata?  Or, are you limited to only the metadata describing the composite resource as a whole?

Is that what you are asking @zhiyuli ?

**Comment by selimnairb** [comment]
@horsburgh @zhiyuli I am thinking that we go for a light weight solution to start with with minimal metadata available for resource files.  Something like what the [resource file list](https://www.hydroshare.org/hsapi/resource/a94af692e2d1442cad3797d23ed7cdf8/file_list/) REST API returns:

```
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "url": "http://www.hydroshare.org/django_irods/download/a94af692e2d1442cad3797d23ed7cdf8/data/contents/Onion.tif",
            "size": 76477563,
            "content_type": "image/tiff"
        }
    ]
}
```
However we should discuss use cases at the all-hands to make sure the simple solution supports key App use cases.  I think that it does.  Also, for some types, e.g. geographic raster and vector, there is nothing to stop the App from inspecting metadata itself once it gets the file in its hands.

**Comment by zhiyuli** [comment]
@horsburgh Yes, that is what I am asking. The whole planned changes here are to make apps more re-usable. Say, an app designed to work with a raster resource should also be able to work with a single raster file in a "composite" resource. The raster resource is metadata-rich (one of selling points of HS), should we make the single raster file in "composite" resource metadata-rich as well? What are the pros and cons of doing or not doing this? What are the costs for hydroshare developers and for HS users? Do we need new plans -- eg., create separate resource for each file and reference them in the "composite" resource just like what current collection does?...A recent request from CyberGIS team asked us to make webapp url more flexible and carry more metadata about the requested resource (file) in query string. We just implemented this in PR-1003. Will the planned changes impact this new-born feature?

@selimnairb I think it is a good start. we could discuss more at the AHM. Thanks

**Comment by selimnairb** [comment]
@zhiyuli I understand the concern.  The metadata that HydroShare provides is important.  However I think there is a place for being able to address light weight objects within resources (i.e. resource files) that have less metadata than the heavy weight objects (i.e. HydroShare resources).  The reason is that we can't hope to model and store, within HydroShare's metadata storage, the complete metadata of every single entity.  Partly this is because our resource types currently require far too much code be written to implement, and also because metadata fields can change, and will do so faster than HydroShare can keep up.  

However, I don't think we don't need "full metadata" to still do useful things with light weight objects (visualize a raster or time series dataset).  Also, there is nothing to stop a specific resource type from having its own metadata store that contains other information about resource files, only a portion of which is reflected in HydroShare's metadata representation (this is how the in-progress RHESSys model instance works).  An App could still be made to be resource type specific (we can create content types that correspond to resource types), in which case the App could use this type-specific metadata as needed.

Let's talk more next week.

**Comment by horsburgh** [comment]
@zhiyuli and @selimnairb - happy to chat about this next week. Given that many content files that could potentially be part of a "Composite" resource actually contain metadata (e.g., a raster has information about extents, etc. within the actual file), there is nothing precluding the App from extracting those metadata from the file itself (even if HydroShare doesn't provide those metadata).  But, you would have to retrieve the file first.  Just a thought.

**Comment by dtarb** [comment]
Lets think of the case of a raster resource.  This holds resource level metadata, such as creator, Abstract etc.  It also contains metadata specific to the raster file, such as the spatial reference and cell size.  These are extracted from the file.  It also contains content specific to the contents of the file that HydroShare has added, such as Variable Name, Units.  (A current use case is we received some groundwater level data from a user as a generic resource and he wrote in the abstract that the variable is Groundwater level and units ft.)  If this was associated with the file the binding would be tighter.  GeoTIFF files do not have a standard model for things like variable name and units.  I suggest that we should consider supporting this file level metadata as optional content associated with resource files in a composite resource.  If it is optional, it should not add additional burden, so we still meet Brian's lightweight requirement.  If we are clever in the coding we can reuse the code for this that is already there for the raster resource type, so should not need too much additional coding burden.  If we did this we would provide the capability for users to have a good place to specify this metadata, and then for apps to look for it and use it where they can take advantage of it (such as displaying the variable name and units in a visualization).

Hopefully next week we can chat about the specifics of some of this and how this idea might generalize to other file types involved.

**Comment by alvacouch** [comment]
There are several things that I notice here: 

1. there are several conflated use cases 

  - allowing apps to infer file types (really, using the MIME extension convention would do a lot here, as well as the 'file' command in Linux that reads file magic numbers independent of MIME type) 
  - giving apps hints -- going beyond simple file type -- that would allow them to more effectively visualize content. 
  - allowing users to understand the provenance of individual files in a resource, independent of app understanding. 

2. these use cases require really different approaches. 

  - app mime type functionality: we can load a library that does this. 
  - an app hint is -- in essence -- a creator-defined or user-defined tag on a file.  
  - the provenance of files -- past file types and tagging -- is a very, very deep issue. Perhaps better handled by gluing resources together as composites. 

I suggest a really lightweight approach to the first two (or really the second one, because we can do the first with a library!): 

  - Allow content providers to tag resource files with app metadata tags that help apps display them.
  -  Keep a list of tags and their meanings, a tag registry.
  - The final arbiter of the tag registry is apps.hydroshre.org 

In other words, 

   - Use general tags. 
   - Allow app developers to design useful tags. 
   - Allow multiple tags per file, to allow different apps to tag them in different ways. 

This would be level one. Level two would be to allow users to suggest file tags independently of the creator. We could -- in essence -- crowdsource the appropriate tags for a resource file.

**Comment by selimnairb** [comment]
@alvacouch Thanks for your thoughts on this.  Your proposed lightweight approach is essentially the same as my proposal, with a bit more detail (e.g. having apps.hydroshare.org be the final arbiter of the tag registry), as well as allowing files to be associated with multiple tags.

**Comment by alvacouch** [comment]
Yes. That is correct. I do think it's important to have a tag registry.

**Comment by selimnairb** [comment]
@alvacouch My initial thought for tags is to create content types from content types supported by existing resource types.  For example, the "geographic raster" content type would be the set of MIME types supported by our version of GDAL.  Do we need to support content types that are not simple sets of MIME types?

**Comment by alvacouch** [comment]
@hyi @dtarb @rayi @danames 
We discussed this on the call today. We understand the basic issue but don't really know what Dan needs. 

But this has been on the back burner too long and I suggest a simple and expedient fix: 

1. Expose the set of MIME types of files in a resource as metadata. This is multi-option checkbox that is set automatically by file type and is not edited by the user. 
2. Let apps register sets of MIME types that they handle (via a multi-checkbox selection). 
3. any overlap between sets means there is some possibility of the app handling files from the resource.

This doesn't solve the general problem: 

1. It solves the problem of visualizing a time series embedded within a model instance, separately ("logical or"). 
2. It does not solve the problem of requiring time series, coverages, etc in a more complex visualization ("logical and").

For the general problem, we would have to allow the app to specify in more detail what kinds of files it absolutely requires, as opposed to the kinds it opportunistically handles. Perhaps we need a flag that switches between "all" and "any" types required?

**Comment by pkdash** [comment]
@alvacouch I did start on this path quite a while ago which you can find in this branch: https://github.com/hydroshare/hydroshare/tree/file-type-web-app
The only difference I see now is that in my implementation a specific app can register only one MIME type which can be easily changed.

**Comment by pkdash** [comment]
Refer to #1477

**Comment by dtarb** [comment]
Closing as I think this has been satisfactorily implemented.  @danames or @alvacouch feel free to reopen, or better yet state a new issue if there are unmet needs from this.
